In [2]:
array=[1,2,3,4,5]
['a','b','c',1,2,3]

['a', 'b', 'c', 1, 2, 3]

In [3]:
import numpy as np

In [4]:
dir(np)

['ALLOW_THREADS',
 'BUFSIZE',
 'CLIP',
 'DataSource',
 'ERR_CALL',
 'ERR_DEFAULT',
 'ERR_IGNORE',
 'ERR_LOG',
 'ERR_PRINT',
 'ERR_RAISE',
 'ERR_WARN',
 'FLOATING_POINT_SUPPORT',
 'FPE_DIVIDEBYZERO',
 'FPE_INVALID',
 'FPE_OVERFLOW',
 'FPE_UNDERFLOW',
 'False_',
 'Inf',
 'Infinity',
 'MAXDIMS',
 'MAY_SHARE_BOUNDS',
 'MAY_SHARE_EXACT',
 'NAN',
 'NINF',
 'NZERO',
 'NaN',
 'PINF',
 'PZERO',
 'RAISE',
 'RankWarning',
 'SHIFT_DIVIDEBYZERO',
 'SHIFT_INVALID',
 'SHIFT_OVERFLOW',
 'SHIFT_UNDERFLOW',
 'ScalarType',
 'True_',
 'UFUNC_BUFSIZE_DEFAULT',
 'UFUNC_PYVALS_NAME',
 'WRAP',
 '_CopyMode',
 '_NoValue',
 '_UFUNC_API',
 '__NUMPY_SETUP__',
 '__all__',
 '__builtins__',
 '__cached__',
 '__config__',
 '__deprecated_attrs__',
 '__dir__',
 '__doc__',
 '__expired_functions__',
 '__file__',
 '__former_attrs__',
 '__future_scalars__',
 '__getattr__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_add_newdoc_ufunc',
 '_builtins',
 '_distributor_init',
 '_financi

In [5]:
list1=[1,2,3,4,5]
array1=np.array(list1)

In [6]:
type(array1)

numpy.ndarray

In [7]:
type(list1)

list

In [8]:
import time
size=1000000
l1=list(range(size))
l2=list(range(size))

start=time.time()
result=[x*y for x,y in zip(l1,l2)]
print("list time:",time.time()-start)

a1=np.arange(size)
a2=np.arange(size)
start=time.time()
result=a1+a2
print("numpy time:",time.time()-start)

list time: 0.09888267517089844
numpy time: 0.09374880790710449


In [9]:
#np.zeros()
zero=np.zeros((20,10))
zero

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])

In [10]:
ones=np.ones((5,6))
ones

array([[1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1.]])

In [11]:
range_array=np.arange(10,150,2)
range_array

array([ 10,  12,  14,  16,  18,  20,  22,  24,  26,  28,  30,  32,  34,
        36,  38,  40,  42,  44,  46,  48,  50,  52,  54,  56,  58,  60,
        62,  64,  66,  68,  70,  72,  74,  76,  78,  80,  82,  84,  86,
        88,  90,  92,  94,  96,  98, 100, 102, 104, 106, 108, 110, 112,
       114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138,
       140, 142, 144, 146, 148])

In [12]:
linspace=np.linspace(0,1,15)
linspace

array([0.        , 0.07142857, 0.14285714, 0.21428571, 0.28571429,
       0.35714286, 0.42857143, 0.5       , 0.57142857, 0.64285714,
       0.71428571, 0.78571429, 0.85714286, 0.92857143, 1.        ])

In [13]:
0.07142857-0.14285714

-0.07142857

In [14]:
linspace=np.linspace(10,150,20)
linspace

array([ 10.        ,  17.36842105,  24.73684211,  32.10526316,
        39.47368421,  46.84210526,  54.21052632,  61.57894737,
        68.94736842,  76.31578947,  83.68421053,  91.05263158,
        98.42105263, 105.78947368, 113.15789474, 120.52631579,
       127.89473684, 135.26315789, 142.63157895, 150.        ])

In [15]:
a=np.array([[1,2,3],[4,5,6]])
print(a)

[[1 2 3]
 [4 5 6]]


In [16]:
a.shape

(2, 3)

In [17]:
linspace.shape

(20,)

In [18]:
a.ndim

2

In [19]:
linspace.ndim

1

In [20]:
a.dtype

dtype('int32')

In [21]:
linspace.dtype

dtype('float64')

In [22]:
linspace.size

20

In [23]:
a.size

6

In [24]:
b=np.array([10,20,30,40,50,60])
b
print("This is b:",b)

This is b: [10 20 30 40 50 60]


In [25]:
b[-1]

60

In [26]:
b[1:4]

array([20, 30, 40])

In [27]:
b[::-1]

array([60, 50, 40, 30, 20, 10])

In [28]:
a

array([[1, 2, 3],
       [4, 5, 6]])

In [29]:
a[0,0]

1

In [30]:
a[0,2]

3

In [31]:
a[:,1] ##all rows of column 1

array([2, 5])

In [32]:
a[1,:] ##all columns of row 1

array([4, 5, 6])

In [33]:
a[1:2]

array([[4, 5, 6]])

In [34]:
c=np.array([
    [[1,2,3],
     [4,5,6]],

    [[7,8,9],[10,11,12]]
])

In [35]:
c.shape

(2, 2, 3)

In [36]:
c[::-1]

array([[[ 7,  8,  9],
        [10, 11, 12]],

       [[ 1,  2,  3],
        [ 4,  5,  6]]])

In [37]:
c[0]

array([[1, 2, 3],
       [4, 5, 6]])

In [38]:
c[0,1,1]

5

In [39]:
c[0,1,0:2]

array([4, 5])

In [40]:
linspace[::-1]

array([150.        , 142.63157895, 135.26315789, 127.89473684,
       120.52631579, 113.15789474, 105.78947368,  98.42105263,
        91.05263158,  83.68421053,  76.31578947,  68.94736842,
        61.57894737,  54.21052632,  46.84210526,  39.47368421,
        32.10526316,  24.73684211,  17.36842105,  10.        ])

In [41]:
a=np.array([1,2,3,4,5])

In [42]:
a[::-1]

array([5, 4, 3, 2, 1])

In [43]:
a=np.array([1,2,3,4,5])
b=np.array([6,7,8,9,10])

In [44]:
a+b

array([ 7,  9, 11, 13, 15])

In [45]:
a-b

array([-5, -5, -5, -5, -5])

In [46]:
a*b

array([ 6, 14, 24, 36, 50])

In [47]:
a**2

array([ 1,  4,  9, 16, 25])

In [48]:
a**b

array([      1,     128,    6561,  262144, 9765625])

In [49]:
numbers=np.array([[1,2,3],[4,5,6]])
numbers

array([[1, 2, 3],
       [4, 5, 6]])

In [50]:
numbers.sum()

21

In [51]:
numbers.mean()

3.5

In [52]:
numbers[0].sum()

6

In [53]:
numbers[1].sum()

15

In [54]:
numbers[0].mean()

2.0

In [55]:
numbers[1].mean()

5.0

In [56]:
numbers[1].min()

4

In [57]:
numbers[0].max()

3

In [58]:
numbers.max()

6

In [59]:
numbers=np.array([[1,2,3],[4,5,6]])
numbers

array([[1, 2, 3],
       [4, 5, 6]])

In [60]:
numbers.sum(axis=0) ##columnwise sum

array([5, 7, 9])

In [61]:
numbers.sum(axis=1) ##rowise sum

array([ 6, 15])

In [62]:
grades=np.array([[85,90,78],
                 [88,92,89],
                 [75,85,89]])

In [63]:
grades[0].max()

90

In [64]:
grades[0].min()

78

In [65]:
grades[0].mean()

84.33333333333333

In [66]:
grades[1].max()

92

In [67]:
grades[1].min()

88

In [68]:
grades[1].mean()

89.66666666666667

In [69]:
grades[2].max()

89

In [70]:
grades[2].min()

75

In [71]:
grades[2].mean()

83.0

#### physics,chemistry,maths
#### min,max,ave,sum of marks of a student , topper in entire class and lowest in entire class and also subjest wise

In [72]:
print("Sum of marks of the topper is:",grades[0].max()+grades[1].max()+grades[2].max())

Sum of marks of the topper is: 271


In [73]:
print("Sum of marks of the lowest is:",grades[0].min()+grades[1].min()+grades[2].min())

Sum of marks of the lowest is: 241


In [74]:
a=np.array([1,2,3])

In [75]:
b=10

In [76]:
a+b

array([11, 12, 13])

In [77]:
A=np.array([
    [10,20,30],
    [40,50,60]])
    #shape(2,3)
B=np.array([[1],[2]]) ## shape(2,1)

In [78]:
A+B

array([[11, 21, 31],
       [42, 52, 62]])

In [79]:
A*B

array([[ 10,  20,  30],
       [ 80, 100, 120]])

In [80]:
check=np.arange(1,20,2)
check

array([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19])

In [81]:
a=check.reshape((2,5))
a

array([[ 1,  3,  5,  7,  9],
       [11, 13, 15, 17, 19]])

In [82]:
a=check.reshape((5,2))
a

array([[ 1,  3],
       [ 5,  7],
       [ 9, 11],
       [13, 15],
       [17, 19]])

In [83]:
a=check.reshape((1,5,2))
a

array([[[ 1,  3],
        [ 5,  7],
        [ 9, 11],
        [13, 15],
        [17, 19]]])

In [84]:
a.shape

(1, 5, 2)

In [85]:
a=check.reshape((1,1,5,2))
a

array([[[[ 1,  3],
         [ 5,  7],
         [ 9, 11],
         [13, 15],
         [17, 19]]]])

In [86]:
a.shape

(1, 1, 5, 2)

#### basic statistics
#### Normalization (scale to 0-1)
#### Topper per subject 
#### Add 5 grace marks to anyone scoring <75 in any subject.Compute overall class average after grace marks.Identify students whose average is >=90 and label them  "Star Performer".

In [87]:
students=np.array(["Alice","Bob","Carol","David","Eva"])
marks=np.array([
    [88,92,85],
    [75,78,80],
    [93,88,90],
    [70,72,68],
    [85,90,87]
    ])
marks

array([[88, 92, 85],
       [75, 78, 80],
       [93, 88, 90],
       [70, 72, 68],
       [85, 90, 87]])

In [88]:
marks.shape

(5, 3)

In [89]:
normalized_data=(marks-marks.min())/(marks.max()-marks.min())
print("The normalized data is:",normalized_data)

The normalized data is: [[0.8  0.96 0.68]
 [0.28 0.4  0.48]
 [1.   0.8  0.88]
 [0.08 0.16 0.  ]
 [0.68 0.88 0.76]]


In [90]:
print("The topper for each subject is",marks.max(axis=0))

The topper for each subject is [93 92 90]


In [91]:
for i in range(5):
    for j in range(3):
        if marks[i][j]<75:
            marks[i][j]+=5
print(marks)

[[88 92 85]
 [75 78 80]
 [93 88 90]
 [75 77 73]
 [85 90 87]]


In [92]:
j=0
for i in marks:
    print("The average of each students",j+1,'is:',i.mean())
    if i.mean()>=90:
        print(students[j],"(Star performer)")
    j+=1

The average of each students 1 is: 88.33333333333333
The average of each students 2 is: 77.66666666666667
The average of each students 3 is: 90.33333333333333
Carol (Star performer)
The average of each students 4 is: 75.0
The average of each students 5 is: 87.33333333333333


# 17-june-2025

## Boolean Index
#### 0-false
#### 1-true

In [93]:
a=np.array([1,5,3,7,9,0,5,4])

In [94]:
a<3

array([ True, False, False, False, False,  True, False, False])

In [95]:
a[a<3]

array([1, 0])

In [96]:
a=np.array([[10,15,20],
            [25,30,35]])

In [97]:
print(a[a>18])

[20 25 30 35]


In [98]:
a[a<20]=0
print(a)

[[ 0  0 20]
 [25 30 35]]


### Matrix multiplication

In [99]:
a=np.array([[1,2,3],[10,20,30]])
b=np.array([[3,6,8],[5,6,2],[5,6,9]])

In [100]:
a.shape

(2, 3)

In [101]:
b.shape

(3, 3)

#### 1.Without using numpy (dot method {dot product of the vector})

In [102]:
a.dot(b) 

array([[ 28,  36,  39],
       [280, 360, 390]])

#### 2. With using numpy (matmul method)

In [103]:
np.matmul(a,b)

array([[ 28,  36,  39],
       [280, 360, 390]])

#### 3. @ method

In [104]:
a@b

array([[ 28,  36,  39],
       [280, 360, 390]])

## Apply along axis

In [105]:
data=np.array([[10,20,30],
               [5,25,25]])

In [106]:
def range_of_row(x):
    return x.max()-x.min()

In [107]:
np.apply_along_axis(range_of_row,axis=0,arr=data)

array([5, 5, 5])

In [108]:
np.apply_along_axis(range_of_row,axis=1,arr=data)

array([20, 20])

In [109]:
data=np.array([[10,23,32],
               [5,25,25]])

In [110]:
def custom(x):
    if x>0 and x<=10:
        return x**3
    elif x>10 and x<=15:
        return x**2
    elif x>15 and x<=20:
        return x**5
    else:
        return x**0.5

In [111]:
data.shape

(2, 3)

### vectorize() (It reduces the time complexity of user defined function, it makes the function in-built function)

In [112]:
vectorized_custom=np.vectorize(custom)

In [113]:
vectorized_custom(data)

array([[1000,    4,    5],
       [ 125,    5,    5]])

In [114]:
arr=np.random.randint(0,100,size=10000000)
print(arr)

[86 47 57 ... 78 86 79]


In [115]:
# method 1: loop(slow)
def square_loop(x):
    return [i*i for i in x]
# method 2: vectorized (fast)
def square_vectorized(x):
    return x**2
%timeit square_loop(arr)
%timeit square_vectorized(arr)

1.57 s ± 300 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
13.7 ms ± 295 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [116]:
# method 1: loop(slow)
def square_loop(x):
    return [i*i for i in x]
# method 2: vectorized (fast)
def square_vectorized(x):
    return x**2
%timeit square_loop(arr)
%timeit square_vectorized(arr)

1.3 s ± 27.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
19 ms ± 1.52 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


## Shallow Copy (Any changes in copied array is reflected in the original array)

In [117]:
a=np.array([1,2,3,4,5])

In [118]:
b=a[1:4]

In [119]:
print("b is :",b)

b is : [2 3 4]


In [120]:
b[0]=322

In [121]:
a

array([  1, 322,   3,   4,   5])

## Deep Copy (Changes not reflected in the original)

In [122]:
c=a[1:4].copy()
print("c is :",c)

c is : [322   3   4]


In [123]:
c[0]=9211

In [124]:
a

array([  1, 322,   3,   4,   5])

## Splitting the array

In [125]:
a=np.array([11,12,17,20,32,45,67,45,89,90,11,99])
a.size

12

#### Splitting equal parts

In [126]:
np.split(a,3)

[array([11, 12, 17, 20]), array([32, 45, 67, 45]), array([89, 90, 11, 99])]

In [127]:
b=np.array([[1,2,3,4],[5,6,7,8]])

In [128]:
np.split(b,2,axis=0)

[array([[1, 2, 3, 4]]), array([[5, 6, 7, 8]])]

In [129]:
np.split(b,2,axis=1)

[array([[1, 2],
        [5, 6]]),
 array([[3, 4],
        [7, 8]])]

In [130]:
np.split(b,4,axis=1)

[array([[1],
        [5]]),
 array([[2],
        [6]]),
 array([[3],
        [7]]),
 array([[4],
        [8]])]

#### splitting unequal parts

In [131]:
np.split(a,[2,4,7])

[array([11, 12]),
 array([17, 20]),
 array([32, 45, 67]),
 array([45, 89, 90, 11, 99])]

In [132]:
np.split(a,[3,5,7,9,11])

[array([11, 12, 17]),
 array([20, 32]),
 array([45, 67]),
 array([45, 89]),
 array([90, 11]),
 array([99])]

### Splitting 2-D array

In [133]:
arr=np.arange(32)
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31])

In [134]:
arr=np.arange(32).reshape(4,8)
arr.shape
arr

array([[ 0,  1,  2,  3,  4,  5,  6,  7],
       [ 8,  9, 10, 11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20, 21, 22, 23],
       [24, 25, 26, 27, 28, 29, 30, 31]])

In [135]:
arr.shape

(4, 8)

In [136]:
np.split(arr,4,axis=0)

[array([[0, 1, 2, 3, 4, 5, 6, 7]]),
 array([[ 8,  9, 10, 11, 12, 13, 14, 15]]),
 array([[16, 17, 18, 19, 20, 21, 22, 23]]),
 array([[24, 25, 26, 27, 28, 29, 30, 31]])]

In [137]:
np.split(arr,4,axis=1)

[array([[ 0,  1],
        [ 8,  9],
        [16, 17],
        [24, 25]]),
 array([[ 2,  3],
        [10, 11],
        [18, 19],
        [26, 27]]),
 array([[ 4,  5],
        [12, 13],
        [20, 21],
        [28, 29]]),
 array([[ 6,  7],
        [14, 15],
        [22, 23],
        [30, 31]])]

In [138]:
np.savetxt("normalized_marks.csv", normalized,delimiter=",",fmt="%.2f",heaader="Marhs,Physics,Chemistry",comments="This my first file saved using numpy")

NameError: name 'normalized' is not defined

# PANDAS

In [ ]:
!pip install pandas

In [139]:
import pandas as pd

In [140]:
marks=pd.Series([88,75,93,70,85],name="Marks")
print(marks)

0    88
1    75
2    93
3    70
4    85
Name: Marks, dtype: int64


In [141]:
data={
    "Name":["Subhendu","Amarendra","Aryan"],
    "Marks":[88,86,93]
}

In [142]:
df=pd.DataFrame(data)
df

Name  Marks
0   Subhendu     88
1  Amarendra     86
2      Aryan     93

In [143]:
data={
    "Name":["Alice","Bob","Carol","David","Eva"],
    "Age":[19,20,19,21,20],
    "Gender":["F","M","F","M","F"],
    "Marks_Math":[88,75,93,70,85],
    "Marks_Science":[92,78,88,72,90],
    "Marks_English":[85,80,90,68,87]
}

In [144]:
df=pd.DataFrame(data)
df

Name  Age Gender  Marks_Math  Marks_Science  Marks_English
0  Alice   19      F          88             92             85
1    Bob   20      M          75             78             80
2  Carol   19      F          93             88             90
3  David   21      M          70             72             68
4    Eva   20      F          85             90             87

In [145]:
df.head()

Name  Age Gender  Marks_Math  Marks_Science  Marks_English
0  Alice   19      F          88             92             85
1    Bob   20      M          75             78             80
2  Carol   19      F          93             88             90
3  David   21      M          70             72             68
4    Eva   20      F          85             90             87

In [146]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Name           5 non-null      object
 1   Age            5 non-null      int64 
 2   Gender         5 non-null      object
 3   Marks_Math     5 non-null      int64 
 4   Marks_Science  5 non-null      int64 
 5   Marks_English  5 non-null      int64 
dtypes: int64(4), object(2)
memory usage: 372.0+ bytes


In [147]:
df.head(2)

Name  Age Gender  Marks_Math  Marks_Science  Marks_English
0  Alice   19      F          88             92             85
1    Bob   20      M          75             78             80

In [148]:
df.head(3)

Name  Age Gender  Marks_Math  Marks_Science  Marks_English
0  Alice   19      F          88             92             85
1    Bob   20      M          75             78             80
2  Carol   19      F          93             88             90

### To read csv files

In [149]:
data=pd.read_csv("customers_data.csv")

FileNotFoundError: [Errno 2] No such file or directory: 'customers_data.csv'

# 18th June 2025

In [ ]:
table=pd.read_html("https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal)")
df=table[0]
print(df)

#### import sqlite3
#### conn=sqlite3.connect("mydb.sqlite")
#### df=pd.read_sql("select*from conn)

### 1)df.head() -> Gives the first entries of the table
### 2)df.tail() -> Gives the last entries of the table
### 3)df.info() -> Gives the information of the table
### 4)df.describe() -> Gives the statistics of the numerical data
### 5)df.shape -> Gives the shape of file
### 6)df.columns -> Gives all the column names
### 7)df['column_name'].value_counts() -> Gives the number of all the duplicate and repeated data in a particular table
### 8)df.index -> Gives start,stop,step of index 
### 9)x=df.set_index("column_name",inplace=True) -> It sets a particular column as index or label and the original data
### 10)df.reset("column_name",inplace=True) -> It resets it original data
### 11)df[['name','age','sex']] -> it gives the particular columns and it should be passed as list if more than one
### 12)df.loc[9995] -> We can access that entry using label
### 13)df.iloc[-5] -> We can access that entry using index
### 14)df.sort_values('column_name',inplace=True) -> Sorts the table according to that column if inplace=true then original data is changed
### 15)df.loc[100:120,'name','age'] -> It gives the range of the data and of the particular columns
### df['age'].mean() -> gives the mean of a particular column
### df['age'].median() -> Gives the median
### df['age'].mode() -> Gives the mode
### df.groupby('column_name').count() -> It groups the data according to a particular column and counts the number for each group
### df.sort_index() -> It sorts the index according to the index

In [150]:
df=pd.DataFrame({
    'Department':['IT','HR','IT','HR','Finance'],
    'Salary':[50000,40000,55000,45000,60000],
    'Bonus':[5000,4000,5500,4200,6000]
})

In [151]:
result=df.groupby("Department").agg({
    "Salary":["mean","min","max"],
    "Bonus":"sum",
    "Department":"count"
})

In [152]:
print(result)

             Salary                Bonus Department
               mean    min    max    sum      count
Department                                         
Finance     60000.0  60000  60000   6000          1
HR          42500.0  40000  45000   8200          2
IT          52500.0  50000  55000  10500          2


In [153]:
df=pd.DataFrame({
    "Name":["Alice","Bob","Carol","Bob"],
    "Math":[90,70,80,90],
    "Science":[85,75,95,75]
})

In [154]:
def total_score(x):
    return x['Math']*0.4+x['Science']*0.6

In [155]:
df['Final_Score']=df.apply(total_score,axis=1)
df

Name  Math  Science  Final_Score
0  Alice    90       85         87.0
1    Bob    70       75         73.0
2  Carol    80       95         89.0
3    Bob    90       75         81.0

In [156]:
def check(x):
    if x>90:
        return 
        x+5
    elif x>80 and x<90:
        return x+3
    else: 
        return x

In [157]:
df['Math'].apply(check)

0    90
1    70
2    80
3    90
Name: Math, dtype: int64

In [158]:
df

Name  Math  Science  Final_Score
0  Alice    90       85         87.0
1    Bob    70       75         73.0
2  Carol    80       95         89.0
3    Bob    90       75         81.0

In [159]:
def check(x):
    return x*x

In [160]:
df.apply(check('Name'))

TypeError: can't multiply sequence by non-int of type 'str'

In [161]:
df.duplicated() ###checks the duplicated data by creating a hashing and checks for collision

0    False
1    False
2    False
3    False
dtype: bool

In [162]:
df.duplicated(subset="Name")

0    False
1    False
2    False
3     True
dtype: bool

In [163]:
df.duplicated(subset="Math")

0    False
1    False
2    False
3     True
dtype: bool

In [164]:
df=pd.DataFrame({
    'Department':['IT','HR','IT','HR','Finance'],
    'Gender':["M","F","M","F","F"],
    'Salary':[50000,40000,55000,45000,60000],
    'Bonus':[5000,4000,5500,4200,6000]
})
df

Department Gender  Salary  Bonus
0         IT      M   50000   5000
1         HR      F   40000   4000
2         IT      M   55000   5500
3         HR      F   45000   4200
4    Finance      F   60000   6000

In [165]:
pivot=pd.pivot_table(df,values="Salary",index="Department",columns="Gender")
print(pivot)

Gender            F        M
Department                  
Finance     60000.0      NaN
HR          42500.0      NaN
IT              NaN  52500.0


In [166]:
df.drop(columns=['Bonus'],inplace=True)

In [167]:
df

Department Gender  Salary
0         IT      M   50000
1         HR      F   40000
2         IT      M   55000
3         HR      F   45000
4    Finance      F   60000

In [168]:
df.rename(columns={'Gender':'Sex'},inplace=True)

In [169]:
df

Department Sex  Salary
0         IT   M   50000
1         HR   F   40000
2         IT   M   55000
3         HR   F   45000
4    Finance   F   60000

In [170]:
df.sample(n=2,random_state=322)

Department Sex  Salary
2         IT   M   55000
4    Finance   F   60000

In [171]:
df


Department Sex  Salary
0         IT   M   50000
1         HR   F   40000
2         IT   M   55000
3         HR   F   45000
4    Finance   F   60000

In [172]:
df.nlargest(2,'Salary')

Department Sex  Salary
4    Finance   F   60000
2         IT   M   55000

In [173]:
df.nlargest(1,'Salary')

Department Sex  Salary
4    Finance   F   60000

In [174]:
df.loc[3,'Sex']

'F'

In [175]:
df.loc[3].at['Department']

'HR'

In [176]:
df.at[3,'Sex'] # both labels

'F'

In [177]:
df.iat[3,1] # one label and one index

'F'

In [178]:
df.query("Salary > 45000")

Department Sex  Salary
0         IT   M   50000
2         IT   M   55000
4    Finance   F   60000

In [179]:
df.query("Salary > 45000 and Department == IT")

UndefinedVariableError: name 'IT' is not defined

In [180]:
c=np.array([
    [[1,2,3],
     [4,5,6]],

    [[7,8,9],[10,11,12]]
])
c

array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]]])

In [181]:
c[::-1,::-1,::-1]

array([[[12, 11, 10],
        [ 9,  8,  7]],

       [[ 6,  5,  4],
        [ 3,  2,  1]]])

In [182]:
df.describe().T

count     mean         std      min      25%      50%      75%  \
Salary    5.0  50000.0  7905.69415  40000.0  45000.0  50000.0  55000.0   

            max  
Salary  60000.0

## Series

In [184]:
sr=pd.Series([1,2,3,4,5])
sr

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [186]:
df=pd.DataFrame({
    "Name":["Alice","Bob","Carol","David"],
    "Age":[25,32,29,40],
    "Income":[50000,60000,55000,70000],
    "Skills":[["Python","SQL"],["Excel"],["Python","ML"],["Java"]],
    "Math":[80,90,85,70],
    "Science":[85,88,90,75]
})
df

Name  Age  Income         Skills  Math  Science
0  Alice   25   50000  [Python, SQL]    80       85
1    Bob   32   60000        [Excel]    90       88
2  Carol   29   55000   [Python, ML]    85       90
3  David   40   70000         [Java]    70       75

In [187]:
#map()
def check(x):
    return x*1.1

In [235]:
df[['Math','Science']].map(check)

Math  Science
0  88.0     93.5
1  99.0     96.8
2  93.5     99.0
3  77.0     82.5

In [236]:
df[['Math','Science']].map(lambda x: x*1.1)

Math  Science
0  88.0     93.5
1  99.0     96.8
2  93.5     99.0
3  77.0     82.5

In [237]:
def add_tax(data):
    data['Income_with_tax']=data['Income']*1.25
    return data

def add_flag(data):
    data["high_income"]=data['Income']>60000
    return data

In [238]:
df.pipe(add_tax).pipe(add_flag)

Name  Age  Income         Skills  Math  Science  Diff_Score  \
0  Alice   25   50000  [Python, SQL]    80       85          -5   
1    Bob   32   60000        [Excel]    90       88           2   
2  Carol   29   55000   [Python, ML]    85       90          -5   
3  David   40   70000         [Java]    70       75          -5   

   Income_with_tax  high_income  
0          62500.0        False  
1          75000.0        False  
2          68750.0        False  
3          87500.0         True

In [239]:
df.explode('Skills')

Name  Age  Income  Skills  Math  Science  Diff_Score  Income_with_tax  \
0  Alice   25   50000  Python    80       85          -5          62500.0   
0  Alice   25   50000     SQL    80       85          -5          62500.0   
1    Bob   32   60000   Excel    90       88           2          75000.0   
2  Carol   29   55000  Python    85       90          -5          68750.0   
2  Carol   29   55000      ML    85       90          -5          68750.0   
3  David   40   70000    Java    70       75          -5          87500.0   

   high_income  
0        False  
0        False  
1        False  
2        False  
2        False  
3         True

In [198]:
df.melt()

variable          value
0              Name          Alice
1              Name            Bob
2              Name          Carol
3              Name          David
4               Age             25
5               Age             32
6               Age             29
7               Age             40
8            Income          50000
9            Income          60000
10           Income          55000
11           Income          70000
12           Skills  [Python, SQL]
13           Skills        [Excel]
14           Skills   [Python, ML]
15           Skills         [Java]
16             Math             80
17             Math             90
18             Math             85
19             Math             70
20          Science             85
21          Science             88
22          Science             90
23          Science             75
24  Income_with_tax        62500.0
25  Income_with_tax        75000.0
26  Income_with_tax        68750.0
27  Income_with_tax        87500.0
28      high_income          False
29      high_income          False
30      high_income          False
31      high_income           True

In [197]:
df.melt('Skills')

Skills         variable    value
0   [Python, SQL]             Name    Alice
1         [Excel]             Name      Bob
2    [Python, ML]             Name    Carol
3          [Java]             Name    David
4   [Python, SQL]              Age       25
5         [Excel]              Age       32
6    [Python, ML]              Age       29
7          [Java]              Age       40
8   [Python, SQL]           Income    50000
9         [Excel]           Income    60000
10   [Python, ML]           Income    55000
11         [Java]           Income    70000
12  [Python, SQL]             Math       80
13        [Excel]             Math       90
14   [Python, ML]             Math       85
15         [Java]             Math       70
16  [Python, SQL]          Science       85
17        [Excel]          Science       88
18   [Python, ML]          Science       90
19         [Java]          Science       75
20  [Python, SQL]  Income_with_tax  62500.0
21        [Excel]  Income_with_tax  75000.0
22   [Python, ML]  Income_with_tax  68750.0
23         [Java]  Income_with_tax  87500.0
24  [Python, SQL]      high_income    False
25        [Excel]      high_income    False
26   [Python, ML]      high_income    False
27         [Java]      high_income     True

In [199]:
## parquet,avro

In [200]:
df

Name  Age  Income         Skills  Math  Science  Income_with_tax  \
0  Alice   25   50000  [Python, SQL]    80       85          62500.0   
1    Bob   32   60000        [Excel]    90       88          75000.0   
2  Carol   29   55000   [Python, ML]    85       90          68750.0   
3  David   40   70000         [Java]    70       75          87500.0   

   high_income  
0        False  
1        False  
2        False  
3         True

In [243]:
#df.set_index("Name")

In [242]:
#st=df.stack[['Math','Science','Income']]

In [241]:
#st.unstack()

#df=pd.DataFrame({
    "Name":["Alice","Bob","Carol","David"],
    "Age":[25,32,29,40],
    "Income":[50000,60000,55000,70000],
    "Skills":[["Python","SQL"],["Excel"],["Python","ML"],["Java"]],
    "Math":[80,90,85,70],
    "Science":[85,88,90,75]
})
df

In [245]:
df.eval("Total_Score = Math + Science",inplace=True)
df

Name  Age  Income         Skills  Math  Science  Diff_Score  \
0  Alice   25   50000  [Python, SQL]    80       85          -5   
1    Bob   32   60000        [Excel]    90       88           2   
2  Carol   29   55000   [Python, ML]    85       90          -5   
3  David   40   70000         [Java]    70       75          -5   

   Income_with_tax  high_income  Total_Score  
0          62500.0        False          165  
1          75000.0        False          178  
2          68750.0        False          175  
3          87500.0         True          145

df=pd.DataFrame({
    "Name":["Alice","Bob","Carol","David"],
    "Age":[25,32,29,40],
    "Income":[50000,60000,55000,70000],
    "Skills":[["Python","SQL"],["Excel"],["Python","ML"],["Java"]],
    "Math":[80,90,85,70],
    "Science":[85,88,90,75]
})
df

In [246]:
#eval() does faster expression
df.eval("Diff_Score = Math - Science",inplace=True)
df

Name  Age  Income         Skills  Math  Science  Diff_Score  \
0  Alice   25   50000  [Python, SQL]    80       85          -5   
1    Bob   32   60000        [Excel]    90       88           2   
2  Carol   29   55000   [Python, ML]    85       90          -5   
3  David   40   70000         [Java]    70       75          -5   

   Income_with_tax  high_income  Total_Score  
0          62500.0        False          165  
1          75000.0        False          178  
2          68750.0        False          175  
3          87500.0         True          145

In [247]:
df.query("Age < 30 and Income >= 55000")

Name  Age  Income        Skills  Math  Science  Diff_Score  \
2  Carol   29   55000  [Python, ML]    85       90          -5   

   Income_with_tax  high_income  Total_Score  
2          68750.0        False          175

In [248]:
df

Name  Age  Income         Skills  Math  Science  Diff_Score  \
0  Alice   25   50000  [Python, SQL]    80       85          -5   
1    Bob   32   60000        [Excel]    90       88           2   
2  Carol   29   55000   [Python, ML]    85       90          -5   
3  David   40   70000         [Java]    70       75          -5   

   Income_with_tax  high_income  Total_Score  
0          62500.0        False          165  
1          75000.0        False          178  
2          68750.0        False          175  
3          87500.0         True          145

In [252]:
df.assign(Bonus = lambda x: x['Income']*0.10)

Name  Age  Income         Skills  Math  Science  Diff_Score  \
0  Alice   25   50000  [Python, SQL]    80       85          -5   
1    Bob   32   60000        [Excel]    90       88           2   
2  Carol   29   55000   [Python, ML]    85       90          -5   
3  David   40   70000         [Java]    70       75          -5   

   Income_with_tax  high_income  Total_Score   Bonus  
0          62500.0        False          165  5000.0  
1          75000.0        False          178  6000.0  
2          68750.0        False          175  5500.0  
3          87500.0         True          145  7000.0

In [250]:
df.assign(Recovery=-100)

Name  Age  Income         Skills  Math  Science  Diff_Score  \
0  Alice   25   50000  [Python, SQL]    80       85          -5   
1    Bob   32   60000        [Excel]    90       88           2   
2  Carol   29   55000   [Python, ML]    85       90          -5   
3  David   40   70000         [Java]    70       75          -5   

   Income_with_tax  high_income  Total_Score  Recovery  
0          62500.0        False          165      -100  
1          75000.0        False          178      -100  
2          68750.0        False          175      -100  
3          87500.0         True          145      -100

In [254]:
df['Name'].unique()

array(['Alice', 'Bob', 'Carol', 'David'], dtype=object)

In [255]:
df['Name'].nunique()

4

In [258]:
df.groupby("Age").count()

Name  Income  Skills  Math  Science  Diff_Score  Income_with_tax  \
Age                                                                     
25      1       1       1     1        1           1                1   
29      1       1       1     1        1           1                1   
32      1       1       1     1        1           1                1   
40      1       1       1     1        1           1                1   

     high_income  Total_Score  
Age                            
25             1            1  
29             1            1  
32             1            1  
40             1            1

In [259]:
df.groupby("Age")["Name"].count()

Age
25    1
29    1
32    1
40    1
Name: Name, dtype: int64

In [260]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             4 non-null      object 
 1   Age              4 non-null      int64  
 2   Income           4 non-null      int64  
 3   Skills           4 non-null      object 
 4   Math             4 non-null      int64  
 5   Science          4 non-null      int64  
 6   Diff_Score       4 non-null      int64  
 7   Income_with_tax  4 non-null      float64
 8   high_income      4 non-null      bool   
 9   Total_Score      4 non-null      int64  
dtypes: bool(1), float64(1), int64(6), object(2)
memory usage: 424.0+ bytes


In [261]:
df['Age'].isna()

0    False
1    False
2    False
3    False
Name: Age, dtype: bool

In [262]:
df.fillna(method='ffill')

C:\Users\ASUS\AppData\Local\Temp\ipykernel_1996\1193302488.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill')


Name  Age  Income         Skills  Math  Science  Diff_Score  \
0  Alice   25   50000  [Python, SQL]    80       85          -5   
1    Bob   32   60000        [Excel]    90       88           2   
2  Carol   29   55000   [Python, ML]    85       90          -5   
3  David   40   70000         [Java]    70       75          -5   

   Income_with_tax  high_income  Total_Score  
0          62500.0        False          165  
1          75000.0        False          178  
2          68750.0        False          175  
3          87500.0         True          145

In [263]:
df.fillna(method='bfill')

C:\Users\ASUS\AppData\Local\Temp\ipykernel_1996\2831856154.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='bfill')


Name  Age  Income         Skills  Math  Science  Diff_Score  \
0  Alice   25   50000  [Python, SQL]    80       85          -5   
1    Bob   32   60000        [Excel]    90       88           2   
2  Carol   29   55000   [Python, ML]    85       90          -5   
3  David   40   70000         [Java]    70       75          -5   

   Income_with_tax  high_income  Total_Score  
0          62500.0        False          165  
1          75000.0        False          178  
2          68750.0        False          175  
3          87500.0         True          145

In [264]:
df.drop(columns='Diff_Score')

Name  Age  Income         Skills  Math  Science  Income_with_tax  \
0  Alice   25   50000  [Python, SQL]    80       85          62500.0   
1    Bob   32   60000        [Excel]    90       88          75000.0   
2  Carol   29   55000   [Python, ML]    85       90          68750.0   
3  David   40   70000         [Java]    70       75          87500.0   

   high_income  Total_Score  
0        False          165  
1        False          178  
2        False          175  
3         True          145

In [274]:
df=pd.DataFrame({
    "Name":["Alice","Bob","Carol","David"],
    "Age":[25,32,None,40],
    "Income":[50000,60000,55000,70000],
    "State":["Montana",None,"LA","New Hampshire"],
    "Sex":["F",None,"M",None],
    "Skills":[["Python","SQL"],["Excel"],["Python","ML"],["Java"]],
    "Math":[80,90,85,70],
    "Science":[85,88,90,75]
})
df

Name   Age  Income          State   Sex         Skills  Math  Science
0  Alice  25.0   50000        Montana     F  [Python, SQL]    80       85
1    Bob  32.0   60000           None  None        [Excel]    90       88
2  Carol   NaN   55000             LA     M   [Python, ML]    85       90
3  David  40.0   70000  New Hampshire  None         [Java]    70       75

In [275]:
df.isna()

Name    Age  Income  State    Sex  Skills   Math  Science
0  False  False   False  False  False   False  False    False
1  False  False   False   True   True   False  False    False
2  False   True   False  False  False   False  False    False
3  False  False   False  False   True   False  False    False

In [276]:
df.isnull().sum()

Name       0
Age        1
Income     0
State      1
Sex        2
Skills     0
Math       0
Science    0
dtype: int64

In [277]:
df.dropna()

Name   Age  Income    State Sex         Skills  Math  Science
0  Alice  25.0   50000  Montana   F  [Python, SQL]    80       85

In [278]:
df.fillna("not provided")

Name           Age  Income          State           Sex         Skills  \
0  Alice          25.0   50000        Montana             F  [Python, SQL]   
1    Bob          32.0   60000   not provided  not provided        [Excel]   
2  Carol  not provided   55000             LA             M   [Python, ML]   
3  David          40.0   70000  New Hampshire  not provided         [Java]   

   Math  Science  
0    80       85  
1    90       88  
2    85       90  
3    70       75

In [281]:
df.fillna({
    'Age':df['Age'].mean(),
    'State':'Unknown',
    'Sex':'3rd Gender'
})

Name        Age  Income          State         Sex         Skills  Math  \
0  Alice  25.000000   50000        Montana           F  [Python, SQL]    80   
1    Bob  32.000000   60000        Unknown  3rd Gender        [Excel]    90   
2  Carol  32.333333   55000             LA           M   [Python, ML]    85   
3  David  40.000000   70000  New Hampshire  3rd Gender         [Java]    70   

   Science  
0       85  
1       88  
2       90  
3       75

In [282]:
df.fillna(0)

Name   Age  Income          State Sex         Skills  Math  Science
0  Alice  25.0   50000        Montana   F  [Python, SQL]    80       85
1    Bob  32.0   60000              0   0        [Excel]    90       88
2  Carol   0.0   55000             LA   M   [Python, ML]    85       90
3  David  40.0   70000  New Hampshire   0         [Java]    70       75

In [285]:
df.fillna(method='ffill')

C:\Users\ASUS\AppData\Local\Temp\ipykernel_1996\1193302488.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df.fillna(method='ffill')


Name   Age  Income          State Sex         Skills  Math  Science
0  Alice  25.0   50000        Montana   F  [Python, SQL]    80       85
1    Bob  32.0   60000        Montana   F        [Excel]    90       88
2  Carol  32.0   55000             LA   M   [Python, ML]    85       90
3  David  40.0   70000  New Hampshire   M         [Java]    70       75

In [284]:
df.bfill()

Name   Age  Income          State   Sex         Skills  Math  Science
0  Alice  25.0   50000        Montana     F  [Python, SQL]    80       85
1    Bob  32.0   60000             LA     M        [Excel]    90       88
2  Carol  40.0   55000             LA     M   [Python, ML]    85       90
3  David  40.0   70000  New Hampshire  None         [Java]    70       75

In [286]:
df.ffill(limit=1)

Name   Age  Income          State Sex         Skills  Math  Science
0  Alice  25.0   50000        Montana   F  [Python, SQL]    80       85
1    Bob  32.0   60000        Montana   F        [Excel]    90       88
2  Carol  32.0   55000             LA   M   [Python, ML]    85       90
3  David  40.0   70000  New Hampshire   M         [Java]    70       75

In [287]:
df['Income'].cumsum()

0     50000
1    110000
2    165000
3    235000
Name: Income, dtype: int64

In [291]:
df['Age'].rolling(window=2).mean()

0     NaN
1    28.5
2     NaN
3     NaN
Name: Age, dtype: float64

In [294]:
df['Sex'].dropna().str.extract(" ")

ValueError: pattern contains no capture groups